Detekt: Add auto-correct functionality for `OrtImportOrder`

MarcelBochtler opened the pull request:
Add the functionality to use Detekt's auto-correct functionality [1] to
fix an incorrect import order or missing empty lines between top level
domains.
To use this, run the `detekt` Gradle task with the `--auto-correct`
flag:
```
./gradlew detekt --auto-correct
```

`auto-correct` can also be useful as a Git pre-commit hook using the
detekt-cli [2]:
```
detekt --plugins /path/to/detekt-rules.jar -c .detekt.yml -i \
  $(git diff --name-only \
    | grep -E '\.kt$|\.kts$' \
    | tr ',\n' ',' \
    | sed '$s/,$/\n/')
```

[1]: https://detekt.dev/docs/introduction/extensions/#autocorrect-property
[2]: https://detekt.dev/cli.html

mnonnenmacher reviewed (commented):
Nice! Could you also update https://github.com/oss-review-toolkit/ort/blob/main/website/docs/development.md#detekt?

sschuberth commented on detekt-rules/src/main/kotlin/OrtImportOrder.kt:
```diff
@@ -66,6 +71,10 @@ class OrtImportOrder(config: Config) : Rule(config) {
         val expectedImportOrder = createExpectedImportOrder(importPaths)
 
         if (importPaths != expectedImportOrder) {
+            if (autoCorrect) {
+                fixImportListOrder(importList, expectedImportOrder)
+            }
```
Nit: This would still for into a single line.

sschuberth commented on detekt-rules/src/main/kotlin/OrtImportOrder.kt:
```diff
@@ -79,6 +88,28 @@ class OrtImportOrder(config: Config) : Rule(config) {
+                importList.add(importDirective)
+                importList.addNewLine()
+            } else {
+                importList.node.addChild(PsiWhiteSpaceImpl("\n"), null)
```
Could `addNewLine()` be used here as well?

sschuberth commented on detekt-rules/src/main/kotlin/OrtImportOrder.kt:
```diff
@@ -79,6 +88,28 @@ class OrtImportOrder(config: Config) : Rule(config) {
+        }
+
+        // Ensure the last empty line is preserved.
+        if (expectedImportOrder.last().isNotEmpty()) {
+            importList.addNewLine()
+        }
```
Nit: This would still for into a single line.

nnobelis commented:
@MarcelBochtler This is really great.
It may look trivial, but I think the ORT import  styling is a pain point for first time contributors.

Does the auto-correct also remove unused imports ?

MarcelBochtler commented:
> Does the auto-correct also remove unused imports ?

We use Detekt's `UnusedImports` rule, which does not support auto-correcting.
Only ktlint formatting rules allow this (or custom rules, like the `OrtImportOrder).

We would therefore need to switch to the `NoUnusedImports` rule, from Ktlint.
Here is the reasoning why we're not already using ktlint rules.
https://github.com/oss-review-toolkit/ort/blob/7f8d3d3540d242a4606f6259f00873fce8203ff8/.detekt.yml#L23-L24

But I did not research if this reasoning is still valid.

sschuberth commented on detekt-rules/src/main/kotlin/OrtImportOrder.kt:
```diff
@@ -68,7 +68,7 @@ class OrtImportOrder(config: Config) : Rule(config) {
             val path = importList.containingKtFile.absolutePath()
             val message = "Imports in file '$path' are not sorted alphabetically or single blank lines are missing " +
-                "between different top-level packages"
+                "between different top-level packages."
```
FYI, while I'm a known fan of properly ending sentences with dots, in this case the dot was deliberately omitted to align with built-in detekt rules, which suffix the message with the name of the rule in square brackets, e.g. `Needless blank line(s) [NoConsecutiveBlankLines]`.

So I'd have a slight preference for dropping this commit. But I'm also ok to keep it.

MarcelBochtler replied:
Dropped the commit, I agree that this message should be in line with other detekt messages.

sschuberth commented:
> Could you also update https://github.com/oss-review-toolkit/ort/blob/main/website/docs/development.md#detekt?

I believe this comment is still open, @MarcelBochtler. An oversight?

MarcelBochtler commented:
> > Could you also update https://github.com/oss-review-toolkit/ort/blob/main/website/docs/development.md#detekt?
> 
> I believe this comment is still open, @MarcelBochtler. An oversight?

I was aware of it ;)
Added the documentation now.

I also dropped the information about a pre-commit hook from the commit message.
I don't think that the `--auto-correct` feature should be used as a pre-commit hook, because the developer should have a look if the changed code is correct, before commiting it.
